python/lsst/cp/pipe/defects.py

```diff
@@ -0,0 +1,645 @@
+            # crop ampImage depending on where the amp lies in the image
+            if self.config.nPixBorderLeftRight:
+                if ampImg.getX0() == 0:
+                    ampImg = ampImg[self.config.nPixBorderLeftRight:, :, afwImage.LOCAL]
+                else:
+                    ampImg = ampImg[:-self.config.nPixBorderLeftRight, :, afwImage.LOCAL]
+            if self.config.nPixBorderUpDown:
+                if ampImg.getY0() == 0:
+                    ampImg = ampImg[:, self.config.nPixBorderUpDown:, afwImage.LOCAL]
+                else:
+                    ampImg = ampImg[:, :-self.config.nPixBorderUpDown, afwImage.LOCAL]
+
+            if self._getNumGoodPixels(ampImg) == 0:  # amp contains no usable pixels
+                continue
+
+            # XXX needs better initialisation, and config options for the MEANCLIP
+            ampImg -= afwMath.makeStatistics(ampImg, afwMath.MEANCLIP, ).getValue()
+
+            mergedSet = None
+            for polarity in polarities:
+                nSig = self.config.nSigmaBright if polarity else self.config.nSigmaDark
+                threshold = afwDetection.createThreshold(nSig, 'stdev', polarity=polarity)
+
+                footprintSet = afwDetection.FootprintSet(ampImg, threshold)
+                if setMask:
+                    footprintSet.setMask(maskedIm.mask, ("DETECTED" if polarity else "DETECTED_NEGATIVE"))
+
+                if mergedSet is None:
+                    mergedSet = footprintSet
+                else:
+                    mergedSet.merge(footprintSet)
+
+            footprintList += mergedSet.getFootprints()
+
+        defects = measAlg.Defects.fromFootprintList(footprintList)
+        return defects
+
+    def _plot(self, dataRef, exp, visit, nSig, defects, imageType):  # pragma: no cover
+        """TODO: Write docstring
+        """
+        afwDisplay.setDefaultBackend("matplotlib")
+        plt.interactive(False)  # seems to need reasserting here
+
+        expCopy = exp.clone()
+
+        dirname = dataRef.getUri(datasetType='cpPipePlotRoot', write=True)
+        if not os.path.exists(dirname):
+            os.makedirs(dirname)
+
+        detNum = dataRef.dataId[self.config.ccdKey]
+
+        if self.config.mode == "MASTER":
+            filename = f"defectPlot_det{detNum}_master-{imageType}_for-exp{visit}.pdf"
+        elif self.config.mode == "VISITS":
+            filename = f"defectPlot_det{detNum}_{imageType}_exp{visit}.pdf"
+
+        filenameFull = os.path.join(dirname, filename)
+        self.log.info(f'Wrote amp histogram to {filenameFull}')
+
+        with warnings.catch_warnings():
+            msg = "Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure."
+            warnings.filterwarnings("ignore", message=msg)
+            with PdfPages(filenameFull) as pdfPages:
+                self._plotAmpHistogram(dataRef, expCopy, visit, nSig)  # NB: modifies exp to scale by expTime
```

Comment:
That comment sounds dangerous! Is that really necessary? The docstring for that method doesn't mention anything like that.

Reply:
Yeah, this was not even really true. I've fixed that up now, sorry about that (like I said, this was never meant to get reviewed 😔 )